src/libraries/System.Runtime.InteropServices.JavaScript/gen/JSImportGenerator/JSExportGenerator.cs

```diff
@@ -209,7 +236,70 @@ private static MarshallingGeneratorFactoryKey<(TargetFramework, Version, JSGener
             return MarshallingGeneratorFactoryKey.Create((env.TargetFramework, env.TargetFrameworkVersion, options), jsGeneratorFactory);
         }
 
-        private static (MemberDeclarationSyntax, ImmutableArray<Diagnostic>) GenerateSource(
+        private static NamespaceDeclarationSyntax GenerateRegSource(
+            ImmutableArray<(StatementSyntax Registration, AttributeListSyntax Attribute)> methods,
+            string assemblyName)
+        {
+            const string generatedNamespace = "System.Runtime.InteropServices.JavaScript";
+            const string initializerClass = "__GeneratedInitializer";
+            const string initializerMethod = "__Register_";
+            const string keepMethod = "__KeepAlive_";
+
+            if (methods.IsEmpty) return NamespaceDeclaration(IdentifierName(generatedNamespace));
+
+            var registerStatements = new List<StatementSyntax>();
+            registerStatements.AddRange(JSExportCodeGenerator.GenerateJSExportArchitectureCheck());
+
+            var attributes = new List<AttributeListSyntax>
+            {
+                //
+            };
+            foreach (var m in methods)
+            {
+                registerStatements.Add(m.Registration);
+                attributes.Add(m.Attribute);
+            }
+
+            FieldDeclarationSyntax field = FieldDeclaration(VariableDeclaration(PredefinedType(Token(SyntaxKind.BoolKeyword)))
+                            .WithVariables(SingletonSeparatedList(
+                                VariableDeclarator(Identifier("initialized")))))
+                            .WithModifiers(TokenList(Token(SyntaxKind.StaticKeyword)));
+
+            MemberDeclarationSyntax method = MethodDeclaration(PredefinedType(Token(SyntaxKind.VoidKeyword)), Identifier(initializerMethod))
+                            .WithAttributeLists(List(attributes))
+                            .WithModifiers(TokenList(new[] { Token(SyntaxKind.StaticKeyword) }))
+                            .WithBody(Block(registerStatements));
+
+            MemberDeclarationSyntax dummyMethod = MethodDeclaration(PredefinedType(Token(SyntaxKind.VoidKeyword)), Identifier(keepMethod))
```

Comment:
Hmm, this is not great. I guess extending DynamicDependencyAttribute to place it at the assembly level could fix that.

@vitek-karas thoughts?

Reply:
Unless we merge this to Net7, we will have to keep `[ModuleInitializer]` anyway. Because otherwise the NET7 runtime would not be able to run nuget packages generated with Net8 code gen.

At let's that's how I understand the `Wasm.Build.Tests.WasmTemplateTests.ConsoleBuildAndRunForSpecificTFM(config: "Debug", extraNewArgs: "-f net7.0") [FAIL]` on CI

I will make that change now and see if that fixed the CI.

Reply:
I'm not used to the source gen code, so this is really hard to read. Could you please paste a sample of the generated code here?

I "think" this is a module initializer method (so something called from the module's .cctor) and it's added only to add a dynamic dependency to something. My question would be how is that something accessed at runtime? This seems like some kind of assembly-level registration mechanism. Maybe it would be better to call some library function from the module initializer to "register itself" instead of using reflection?